core/shared/src/main/scala/scodec/bits/ByteVector.scala

```diff
@@ -1720,6 +1753,53 @@ object ByteVector {
   def fromValidBin(str: String, alphabet: Bases.BinaryAlphabet = Bases.Alphabets.Binary): ByteVector =
     fromBinDescriptive(str, alphabet).fold(msg => throw new IllegalArgumentException(msg), identity)
 
+  /**
+    * Constructs a `ByteVector` from a base 58 string or returns an error message if the string is not valid base 58.
+    * It is similar to Base64 but has been modified to avoid both non-alphanumeric characters and letters which might look ambiguous when printed.
+    * It is therefore designed for human users who manually enter the data, copying from some visual source
+    * Compared to Base64, the following similar-looking letters are omitted: 0 (zero), O (capital o), I (capital i) and l (lower case L)
+    * as well as the non-alphanumeric characters + (plus) and / (slash).
+    * The actual order of letters in the alphabet depends on the application, the default order is the same used in Bitcoin
+    * An empty input string results in an empty ByteVector.
+    * The string may contain whitespace characters which are ignored.
+    * @group base
+    */
+  def fromBase58Descriptive(str: String, alphabet: Bases.Alphabet = Bases.Alphabets.Base58): Either[String, ByteVector] = {
+    val zeroLength = str.takeWhile(_ == '1').length
+    val zeroes = ByteVector.fill(zeroLength.toLong)('0')
+    val trim = str.splitAt(zeroLength)._2.toList
+    val RADIX = BigInt(58L)
+    val decoded = trim.foldLeft(BigInt(0)){ (a, c) =>
+      try {
+        a * RADIX + BigInt(alphabet.toIndex(c))
+      } catch {
+        case e: IllegalArgumentException =>
+          val idx = trim.takeWhile(_ != c).length
+          return Left(s"Invalid base 58 character '$c' at index $idx")
+      }
+    }
+    if (trim.isEmpty) Right(zeroes) else Right(zeroes ++ ByteVector(decoded.toByteArray.dropWhile(_ == 0)))
```

Comment:
I changed to be just `if (trim.isEmpty) Right(zeroes) else Right(zeroes ++ ByteVector(decoded))` because i assume the alphabet will always be in the same order so 1 will always be the first character. But does scodec-bits encourage for users to have their own alphabets?

Reply:
nvm i notice is due to toByteArray sometimes adding a 0

Reply:
`toByteArray` also preapends a '00' in some cases.